### PR TITLE
[#1140] adds warning when the target has no target properties

### PR DESCRIPTION
**Pull request by zyberzebra**
Fixes: #1140 
TODO:

- [x] should no accessible properties also lead to the warning? 
   - [x] I guess so
- [x] remove reportErrorForUnmappedTargetPropertiesIfRequired, I guess?
   - realized unmappedTarget is a different case.
- [x] Any other possibilities or edge cases to test?

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java [hunk 1]
```diff
@@ -232,6 +232,14 @@ else if ( !method.isUpdateMethod() ) {
             // determine accessors
             Map<String, Accessor> accessors = resultTypeToMap.getPropertyWriteAccessors( cms );
             this.targetProperties = new LinkedHashSet<>( accessors.keySet() );
+            if ( targetProperties.isEmpty() ) {
```
Comment (zyberzebra) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:235]: - [ ] Do we have a Validation Framework for stuff like that?

Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:235]: I guess the validation framework is all those ifs in the build method haha

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 2]
```diff
@@ -225,8 +229,10 @@ public enum Message {
     ANNOTATE_WITH_ENUM_CLASS_NOT_DEFINED( "enumClass needs to be defined when using enums." ),
     ANNOTATE_WITH_ENUMS_NOT_DEFINED( "enums needs to be defined when using enumClass." ),
     ANNOTATE_WITH_ANNOTATION_IS_NOT_REPEATABLE( "Annotation \"%s\" is not repeatable." ),
-    ANNOTATE_WITH_DUPLICATE( "Annotation \"%s\" is already present with the same elements configuration.", Diagnostic.Kind.WARNING ),
-    ;
+    ANNOTATE_WITH_DUPLICATE(
+        "Annotation \"%s\" is already present with the same elements configuration.",
+        Diagnostic.Kind.WARNING
+    );
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:235]: Please do not commit unrelated style changes. Can you please remove this change?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 3]
```diff
@@ -92,6 +92,10 @@ public enum Message {
     PROPERTYMAPPING_CANNOT_DETERMINE_SOURCE_PARAMETER_FROM_TARGET("No property named \"%s\" exists in source parameter(s). Please define the source explicitly."),
     PROPERTYMAPPING_NO_SUITABLE_COLLECTION_OR_MAP_CONSTRUCTOR( "%s does not have an accessible copy or no-args constructor." ),
     PROPERTYMAPPING_EXPRESSION_AND_CONDITION_QUALIFIED_BY_NAME_BOTH_DEFINED( "Expression and condition qualified by name are both defined in @Mapping, either define an expression or a condition qualified by name." ),
+    PROPERTYMAPPING_TARGET_HAS_NO_TARGET_PROPERTIES(
+        "Target \"%s\" has no target properties, targeted by this mapping method: \"%s\"",
+        Diagnostic.Kind.WARNING
+    ),
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:98]: Please put this on one line in order to be consistent with how it is for the other messages.

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1140/AirplaneWithNoAccessors.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package org.mapstruct.ap.test.bugs._1140;
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1140/AirplaneWithNoAccessors.java:1]: I wouldn't call this a bug. I think some other test package would be a better fit.

File: processor/src/test/java/org/mapstruct/ap/test/bugs/_1140/JarMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.compilation.annotation.CompilationResult;
+import org.mapstruct.ap.testutil.compilation.annotation.Diagnostic;
+import org.mapstruct.ap.testutil.compilation.annotation.ExpectedCompilationOutcome;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+class JarMapperTest {
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/bugs/_1140/JarMapperTest.java:12]: The assertions in the test are irrelevant. What we want to test here is only the messages we have during compilation. So only `@ExpectedCompilationOutcome` for this

**Review by filiphr — COMMENTED** [review]
Thanks for your work on this @zyberzebra. I've left some comments.

**Comment by zyberzebra** [comment]
@filiphr I'm not 100% what the correct workflow is. Should I request reviews when I'm ready or not?

**Comment by filiphr** [comment]
The workflow is me trying to get time and going through all the emails and updated PRs 😀. I usually go through PRs that have been updated as well

**Comment by zyberzebra** [comment]
> The workflow is me trying to get time and going through all the emails and updated PRs 😀. I usually go through PRs that have been updated as well 

Thanks for the clarification ! I just scrolled through some other pull request here and saw that some request reviews and because it's my first time working open source I was unsure. ;)
At work I just ping my mates 😁

**Comment by zyberzebra** [comment]
@filiphr Sorry for the ping. Just thought I give you a ping, because my PRs are probably way down on the pile now. 

With more time passing, it gets harder for me to remember what we fixed. Which is not a problem, put it might take longer to for me to respond if you come around for a review. :)

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 6]
```diff
@@ -92,6 +92,7 @@ public enum Message {
     PROPERTYMAPPING_CANNOT_DETERMINE_SOURCE_PARAMETER_FROM_TARGET("No property named \"%s\" exists in source parameter(s). Please define the source explicitly."),
     PROPERTYMAPPING_NO_SUITABLE_COLLECTION_OR_MAP_CONSTRUCTOR( "%s does not have an accessible copy or no-args constructor." ),
     PROPERTYMAPPING_EXPRESSION_AND_CONDITION_QUALIFIED_BY_NAME_BOTH_DEFINED( "Expression and condition qualified by name are both defined in @Mapping, either define an expression or a condition qualified by name." ),
+    PROPERTYMAPPING_TARGET_HAS_NO_TARGET_PROPERTIES( "Target \"%s\" has no target properties, targeted by this mapping method: \"%s\"", Diagnostic.Kind.WARNING ),
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:95]: Do we really need the mapping method here? The line on which we report the warning is the method itself so it should not be needed to also put the method here.

File: processor/src/test/java/org/mapstruct/ap/test/emptytarget/JarMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import org.mapstruct.ap.testutil.compilation.annotation.Diagnostic;
+import org.mapstruct.ap.testutil.compilation.annotation.ExpectedCompilationOutcome;
+
+
+class JarMapperTest {
+
+    @ProcessorTest
+    @ExpectedCompilationOutcome(value = CompilationResult.SUCCEEDED,
+        diagnostics = {
+            @Diagnostic(type = JarToJarMapper.class,
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/emptytarget/JarMapperTest.java:20]: We should also add the `line = 16` on the diagnostic as well

File: processor/src/test/java/org/mapstruct/ap/test/emptytarget/JarMapperTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        })
+    @WithClasses({ FilledJar.class, EmptyJar.class, JarToJarMapper.class })
+    void targetHasNoProperties() {
+    }
+
+
+    @ProcessorTest
+    @ExpectedCompilationOutcome(value = CompilationResult.SUCCEEDED,
+        diagnostics = {
+            @Diagnostic(type = JarToAirplaneMapper.class,
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/emptytarget/JarMapperTest.java:32]: We should also add the `line = 16` on the diagnostic as well

**Review by filiphr — CHANGES_REQUESTED** [review]
Sorry for the waiting @zyberzebra. I've done another pass here. Have a look at my comments

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java [hunk 9]
```diff
@@ -232,6 +232,9 @@ else if ( !method.isUpdateMethod() ) {
...
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:235]: I missed this in my previous review. Looking at this if the target does not have setters but has constructor arguments then reporting it here before the constructor has been processed is incorrect. Shouldn't we move this after `this.targetProperties` is no longer updated (e.g. on line 257 it is populated)? See the adjusted tests.

**Review by filiphr — CHANGES_REQUESTED** [review]
Good work @zyberzebra. I've done some small polishing and added a test case for the comment I left.

Have a look at it and let me know what you think

File: processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java (hunk 9 above)
Comment (zyberzebra) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java:235]: You are right! 
I think we also have to come up with a solution for the newly added warning, which leads to a lot of existing Tests failing.

Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java]: Let me see what kind of tests are failing. Mapping into an empty object is anyways a bit strange. Perhaps we need to adjust those test cases.
